nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp

```diff
@@ -0,0 +1,570 @@
+  if (!transformPose(costmap_ros_->getGlobalFrameID(), carrot_pose, carrot_in_odom))
+  {
+    RCLCPP_ERROR(logger_, "Unable to get carrot pose in odom frame, failing collision check!");
+    return true;
+  }
+
+  // check current point and carrot point are OK, most likely ones to be in collision
+  if (inCollision(carrot_in_odom.pose.position.x, carrot_in_odom.pose.position.y) ||
+    inCollision(robot_pose.pose.position.x, robot_pose.pose.position.y))
+  {
+    return true;
+  }
+
+  // debug messages
+  nav_msgs::msg::Path arc_pts_msg;
+  arc_pts_msg.header.frame_id = costmap_ros_->getGlobalFrameID();
+  arc_pts_msg.header.stamp = robot_pose.header.stamp;
+  geometry_msgs::msg::PoseStamped pose_msg;
+  pose_msg.header.frame_id = arc_pts_msg.header.frame_id;
+  pose_msg.header.stamp = arc_pts_msg.header.stamp;
+
+  // Using curvature (k = 1 / R) and carrot distance (chord on circle R), project the command
+  const double chord_len = hypot(carrot_pose.pose.position.x, carrot_pose.pose.position.y);
+
+  // Find the number of increments by finding the arc length of the chord on circle
+  const double r = fabs(1.0 / curvature);
+  const double alpha = 2.0 * asin(chord_len / (2 * r)); // central angle
+  const double arc_len = alpha * r;
+  const double delta_dist = costmap_->getResolution();
+  const unsigned int num_pts = static_cast<unsigned int>(ceil(arc_len / delta_dist));
+  const double projection_time = costmap_->getResolution() / linear_vel;
+
+  geometry_msgs::msg::Pose2D curr_pose;
+  curr_pose.x = robot_pose.pose.position.x;
+  curr_pose.y = robot_pose.pose.position.y;
+  curr_pose.theta = tf2::getYaw(robot_pose.pose.orientation);
+
+  for (unsigned int i = 1; i < num_pts; i++) {
+    // only forward simulate within time requested
+    if (i * projection_time > max_allowed_time_to_collision_) {
+      break;
+    }
+
+    // apply velocity at curr_pose over distance delta_dist
+    curr_pose.x += projection_time * (linear_vel * cos(curr_pose.theta));
+    curr_pose.y += projection_time * (linear_vel * sin(curr_pose.theta));
+    curr_pose.theta += projection_time * angular_vel;
+
+    // store it for visualization
+    pose_msg.pose.position.x = curr_pose.x;
+    pose_msg.pose.position.y = curr_pose.y;
+    pose_msg.pose.position.z = 0.01;
+    arc_pts_msg.poses.push_back(pose_msg);
+
+    // check for collision at this point
+    if (inCollision(curr_pose.x, curr_pose.y)) {
+      carrot_arc_pub_->publish(arc_pts_msg);
+      return true;
+    }
+  }
+
+  carrot_arc_pub_->publish(arc_pts_msg);
+
+  return false;
+}
+
+bool RegulatedPurePursuitController::inCollision(const double & x, const double & y)
+{
+  unsigned int mx, my;
+  costmap_->worldToMap(x, y, mx, my);
+
+  unsigned char cost = costmap_->getCost(mx, my);
+
+  if (costmap_ros_->getLayeredCostmap()->isTrackingUnknown()) {
+    return cost >= INSCRIBED_INFLATED_OBSTACLE && cost != NO_INFORMATION;
+  } else {
+    return cost >= INSCRIBED_INFLATED_OBSTACLE;
```

Comment:
Could here be a situation when local costmap's default value another than `NO_INFORMATION`, but after some costmap layers `NO_INFORMATION` still might appear on costmaps? E.g. there is no guarantee that some algorithms or costmap filters in the future might not emit such costmaps. In that case we will consider no-information as a collision. In my opinion, `cost >= INSCRIBED_INFLATED_OBSTACLE && cost != NO_INFORMATION` condition is sufficient for both cases. If-statement in this case is unnecessary.

Reply:
Sure, but I'm not sure what that has to do with this - unknown is a special class type that needs to be handled uniquely from any other cost value because of the fact that sometimes you track it and other times you don't. Its the general understanding in all of the plugins that if you track it then its no-go space because you're not confident about anything about it (like if there's a drop off for stairs). The default value of the costmap isn't necessarily related to this check. The default could be 100 for some reason, but we still need to handle no information as a special case.

Reply:
Sorry, also rereading this - was there a specific action I can take here?

Reply:
OK, thanks for clarification. That sounds reasonable. However, one point I still not confident about. You've mentioned that:
 > Its the general understanding in all of the plugins that if you track it then its no-go space

In other words, if we are tracking unknowns, the areas with unknown values will be treated as no-go, right? In this case we should return `true` if `cost == NO_INFROMATION` for `isTrackingUnknown()`. In the code, the situation is seems to be opposite: currently we are consider unknown places as no-go if `isTrackingUnknown()` is false.

> I could see where you think that.

I've just checked the code implementation of `isTrackingUnknown()`: it return `true` if costmap's default value is `NO_INFORMATION`, otherwise - `false`. Therefore, I've pointed my interest on this statement. How costmap's default value could affect a moving through unknown area decisions?

> Sorry, also rereading this - was there a specific action I can take here?

After reading a clarification that you've gave, if-statement here looks to have a reason. However, let's make it clear for the first question in this comment. When should we avoid unknowns: if it is tracked or if it is not tracked?

Reply:
>  the areas with unknown values will be treated as no-go, right?

Not necessarily. Again, that's to do with if we track it or not. Different algorithms will do different things, that's why I check it

Reply:
We could add a param whether unknown space is traversible or not, but it seems redundant since the costmap provides us that implicitly

Reply:
Agree with last argument. This does not make sense.